### PR TITLE
issue #1351 - Check SearchParameter search restrictions

### PR DESCRIPTION
**Pull request by tbieste**
This update enables checking of SearchParameter search restrictions and updates the FHIR server config file with a resources group that will provide configuration for additional search restrictions. Checking of required search parameters, includes, and revincludes will be in an upcoming commit.

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 1]
```diff
@@ -800,7 +810,80 @@ public static FHIRSearchContext parseQueryParameters(Class<?> resourceType,
...
+     * @param searchParameter
+     *            the SearchParameter resource
+     * @param queryParameters
+     *            the query parameters to check
+     * @throws FHIRSearchException
+     *             if a search restriction is found that is not followed
+     */
+    private static void checkSearchParameterRestrictions(String parameterCode, SearchParameter searchParameter, List<QueryParameter> queryParameters)
+        throws FHIRSearchException {
+
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:829]: I see all the failures.  We may want to add a configuration to turn this feature on/off.

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:829]: Hopefully not. I was using the Draft Pull Request to run the e2e tests before this is ready for review, and noticed a fix I needed to make.  We'll see if that takes care of the errors.

**Comment by tbieste** [comment]
@lmsurpre An example config with the new fields is fhir-search/src/test/resources/config/tenant7/fhir-server-config.json
I'll be updating it as I add some more unit tests in an upcoming commit, but I should give you a good example to start with.

File: fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java [hunk 2]
```diff
@@ -36,8 +36,13 @@
     public static final String PROPERTY_CAPABILITY_STATEMENT_CACHE = "fhirServer/core/capabilityStatementCacheTimeout";
     public static final String PROPERTY_EXTENDED_CODEABLE_CONCEPT_VALIDATION = "fhirServer/core/extendedCodeableConceptValidation";
 
-    public static final String PROPERTY_SEARCH_PARAMETER_FILTER = "fhirServer/searchParameterFilter";
-
+    // Resources properties
+    public static final String PROPERTY_RESOURCES = "fhirServer/resources";
+    public static final String PROPERTY_FIELD_RESOURCES_INCLUDE_OMITTED = "includeOmittedResourceTypes";
```
Comment (lmsurpre) [inline fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:41]: Paul wasn't a big fan of this property name, so we just had a chat and settled on `supportOmittedResourceTypes`:
```suggestion
    public static final String PROPERTY_FIELD_RESOURCES_SUPPORT_OMITTED = "supportOmittedResourceTypes";
```

Comment (JohnTimm) [reply fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:41]: What about `open`as the property name? Where `open=true` allows inclusion of resources not explicitly defined in the configuration.

Comment (lmsurpre) [reply fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:41]: Talked with Paul and we both like it.
```suggestion
    public static final String PROPERTY_FIELD_RESOURCES_OPEN = "open";
```

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 3]
```diff
@@ -293,51 +295,72 @@ public static void init() {
...
+        
+        // Retrieve the "resources" config property group.
+        PropertyGroup rsrcsGroup = FHIRConfigHelper.getPropertyGroup(FHIRConfiguration.PROPERTY_RESOURCES);
+        if (rsrcsGroup != null) {
+            List<PropertyEntry> rsrcsEntries = rsrcsGroup.getProperties();            
+            if (rsrcsEntries != null && !rsrcsEntries.isEmpty()) {
+                for (PropertyEntry rsrcsEntry : rsrcsEntries) {
+                    
+                    // Check special property for including omitted resource types
+                    if (FHIRConfiguration.PROPERTY_FIELD_RESOURCES_INCLUDE_OMITTED.equals(rsrcsEntry.getName())) {
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:314]:
```suggestion
                    if (FHIRConfiguration.PROPERTY_FIELD_RESOURCES_OPEN.equals(rsrcsEntry.getName())) {
```

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 4]
```diff
@@ -293,51 +295,72 @@ public static void init() {
...
-        // If we have a non-empty set of filter rules, then walk through them and populate our map.
-        if (ruleEntries != null && !ruleEntries.isEmpty()) {
-            for (PropertyEntry ruleEntry : ruleEntries) {
-                String resourceType = ruleEntry.getName();
-
-                // Make sure the value is a List<String>.
-                if (ruleEntry.getValue() instanceof List<?>) {
-                    for (Object listMember : (List<?>) ruleEntry.getValue()) {
-                        if (!(listMember instanceof String)) {
+        boolean includeOmittedRsrcTypes = true;
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:304]:
```suggestion
        boolean supportOmittedRsrcTypes = true;
```

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 5]
```diff
@@ -293,51 +295,72 @@ public static void init() {
...
+        PropertyGroup rsrcsGroup = FHIRConfigHelper.getPropertyGroup(FHIRConfiguration.PROPERTY_RESOURCES);
+        if (rsrcsGroup != null) {
+            List<PropertyEntry> rsrcsEntries = rsrcsGroup.getProperties();            
+            if (rsrcsEntries != null && !rsrcsEntries.isEmpty()) {
+                for (PropertyEntry rsrcsEntry : rsrcsEntries) {
+                    
+                    // Check special property for including omitted resource types
+                    if (FHIRConfiguration.PROPERTY_FIELD_RESOURCES_INCLUDE_OMITTED.equals(rsrcsEntry.getName())) {
+                        if (rsrcsEntry.getValue() instanceof Boolean) {
+                            includeOmittedRsrcTypes = (Boolean) rsrcsEntry.getValue();
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:316]:
```suggestion
                            supportOmittedRsrcTypes = (Boolean) rsrcsEntry.getValue();
```

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 6]
```diff
@@ -293,51 +295,72 @@ public static void init() {
...
-        } else {
-            // The current tenant doesn't have any filter rules defined, so
-            // we'll just fabricate one that includes all search parameters:
-            // <pre>{ "*": ["*"] }</pre>
-            List<String> list = new ArrayList<>();
-            list.add(SearchConstants.WILDCARD);
-            result.put(SearchConstants.WILDCARD, list);
         }
+
+        if (includeOmittedRsrcTypes) {
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:356]:
```suggestion
        if (supportOmittedRsrcTypes) {
```

File: fhir-search/src/test/resources/config/tenant1/fhir-server-config.json [hunk 7]
```diff
@@ -1,23 +1,41 @@
...
-				"active",
-				"address",
-				"birthdate",
-				"name"
-			],
-			"*": [
-				"*"
-			]
+		"resources": {
+			"includeOmittedResourceTypes": true,
```
Comment (lmsurpre) [inline fhir-search/src/test/resources/config/tenant1/fhir-server-config.json:5]:
```suggestion
			"open": true,
```

File: fhir-search/src/test/resources/config/tenant4/fhir-server-config.json [hunk 8]
```diff
@@ -1,23 +1,41 @@
...
-				"active",
-				"address",
-				"birthdate",
-				"name"
-			],
-			"*": [
-				"*"
-			]
+		"resources": {
+			"includeOmittedResourceTypes": true,
```
Comment (lmsurpre) [inline fhir-search/src/test/resources/config/tenant4/fhir-server-config.json:5]:
```suggestion
			"open": true,
```

File: fhir-search/src/test/resources/config/tenant5/fhir-server-config.json [hunk 9]
```diff
@@ -1,23 +1,41 @@
...
-				"active",
-				"address",
-				"birthdate",
-				"name"
-			],
-			"*": [
-				"*"
-			]
+		"resources": {
+			"includeOmittedResourceTypes": true,
```
Comment (lmsurpre) [inline fhir-search/src/test/resources/config/tenant5/fhir-server-config.json:5]:
```suggestion
			"open": true,
```

File: fhir-search/src/test/resources/config/tenant7/fhir-server-config.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
+{
+	"__comment": "FHIR Server configuration",
+	"fhirServer": {
+		"resources": {
+			"includeOmittedResourceTypes": true,
```
Comment (lmsurpre) [inline fhir-search/src/test/resources/config/tenant7/fhir-server-config.json:5]:
```suggestion
			"open": true,
```

File: fhir-server/liberty-config-tenants/config/tenant1/fhir-server-config.json [hunk 11]
```diff
@@ -1,13 +1,34 @@
...
 	"fhirServer": {
-		"searchParameterFilter": {
-			"Observation": ["subject",
-			"patient",
-			"value-quantity",
-			"component-value-quantity"],
-			"Resource": ["_id"],
-			"*": ["*"]
+		"resources": {
+			"includeOmittedResourceTypes": true,
```
Comment (lmsurpre) [inline fhir-server/liberty-config-tenants/config/tenant1/fhir-server-config.json:5]:
```suggestion
			"open": true,
```

File: fhir-server/liberty-config-tenants/config/tenant1/fhir-server-config.json [hunk 12]
```diff
@@ -1,13 +1,34 @@
...
+			"Observation": {
+				"searchParameters": {
+					"_id": {
+						"url": "http://hl7.org/fhir/SearchParameter/Resource-id"
+					},
```
Comment (lmsurpre) [inline fhir-server/liberty-config-tenants/config/tenant1/fhir-server-config.json:10]: shouldn't be needed (it should inheret from Resource), but should be fine...

**Review by lmsurpre — APPROVED** [review]
LGTM

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 13]
```diff
@@ -293,51 +295,72 @@ public static void init() {
...
-            // we'll just fabricate one that includes all search parameters:
-            // <pre>{ "*": ["*"] }</pre>
-            List<String> list = new ArrayList<>();
-            list.add(SearchConstants.WILDCARD);
-            result.put(SearchConstants.WILDCARD, list);
         }
+
+        if (supportOmittedRsrcTypes) {
+            // All other resource types include all search parameters
+            result.put(SearchConstants.WILDCARD, Collections.singletonList(SearchConstants.WILDCARD));
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:358]: would it be helpful (perf-wise) to have this singletonList in a constant?

**Review by prb112 — APPROVED** [review]
LGTM

**Comment by lmsurpre** [comment]
we still need to update the docs to reflect the new config changes but I will do that as part of #1449

**Pull request merged**